packages/@lwc/synthetic-shadow/src/shared/handled-events.ts

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+const currentlyHandledEvents: Set<Event> = new Set();
```

Comment:
@caridy We had discussed about using a global flag to indicate that the engine is currently invoking a wrapped listener. 

IMO, the Set usage to track events is better than the global flag. Consider this scenario:
1. The engine sets the flag and invokes a wrapped listener
2. The listener fires off another DOM event which has a global listener(which expects the un-retargeted event)
3. When the global listener accesses the event's target, the global flag will still be set and the global listener will get the retargeted event.

```
// Listener is added before lwc engine is loaded
document.addEventListener('EventA', (event) => {
    event.target; // Will be retargeted
});

// listener added after lwc engine is loaded
export default MyComponent extends LightningElement {
    clickHandler() {
           this.dispatchEvent(new CustomEvent('EventA', {bubbles: true, composed: true})); // This will synchronously invoke the listener on the document.
    }
}

```

Reply:
Yeah, you can have one or more events with overlapping lifecycles at any given time.

Reply:
I do agree that the set is nicer, but... jejeje... there is always a but... the problem is circular listeners, e.g.: if the same listener is triggered twice, what happen then? The closure is safer IMO.

Reply:
@caridy What's an example of a listener being triggered twice?

Reply:
```
el.addEventListener('foo', monitorFooEvent);
el.addEventListener('foo', dispatchFooInAnotherElem);
anotherElm.addEventListener('foo', monitorFooEvent)
```

think about monitoring, or any other event that gets applied to multiple elements.

Reply:
Right, the relationship between events and listeners is many-to-many. But I'm not sure what exactly the bug would be. For instance:

```js
el.addEventListener('click', listener)
el.addEventListener('keydown', listener)
```

Let's say that the user focuses the element and presses <kbd>Enter</kbd>. These events would both fire on `keydown`, but we would still handle them sequentially, so I don't see what the bug is.

Reply:
right, not there... the problem is when a listener dispatches another event, and that new event somehow, triggers a listener that is already in the set, that's when this is broken, because once the second installment removes the listener from the set, once it gets back the control, if there is any other pending code to be executed, accessing e.target at that point will not get retargeted, because the current listener is not longer in the set. I understand that this is a rather edge case, but still possible.

Reply:
@caridy We are tracking events, not listeners, and it's not possible for more than one listener to simultaneously process a particular event, so I don't believe we need to worry about the scenario you described.

Reply:
oh shit! jajajaja! I missed that little (but very important) detail... all good!